contracts/multisig-prover/src/encoding/bcs.rs

```diff
@@ -0,0 +1,284 @@
+use std::iter;
+
+use axelar_wasm_std::hash::Hash;
+use cosmwasm_std::HexBinary;
+use error_stack::{Result, ResultExt};
+use multisig::msg::SignerWithSig;
+use multisig::verifier_set::VerifierSet;
+use sha3::{Digest, Keccak256};
+use sui_gateway::gateway::{
+    ExecuteData, Message, MessageToSign, Proof, WeightedSigners, COMMAND_TYPE_APPROVE_MESSAGES,
+    COMMAND_TYPE_ROTATE_SIGNERS,
+};
+
+use crate::encoding::{to_recoverable, Encoder};
+use crate::error::ContractError;
+use crate::payload::Payload;
+
+fn encode_payload(payload: &Payload) -> Result<Vec<u8>, ContractError> {
+    let encoded: Vec<u8> = match payload {
+        Payload::Messages(messages) => bcs::to_bytes(
+            &messages
+                .iter()
+                .map(Message::try_from)
+                .collect::<Result<Vec<_>, _>>()
+                .change_context(ContractError::InvalidMessage)?,
+        )
+        .expect("failed to serialize messages"),
+        Payload::VerifierSet(verifier_set) => {
+            bcs::to_bytes(&WeightedSigners::from(verifier_set.clone()))
+                .expect("failed to weighted signers")
+        }
+    };
+
+    Ok(encoded)
+}
+
+pub fn payload_digest(
+    domain_separator: &Hash,
+    verifier_set: &VerifierSet,
+    payload: &Payload,
+) -> Result<Hash, ContractError> {
+    let command_type = match payload {
+        Payload::Messages(_) => COMMAND_TYPE_APPROVE_MESSAGES,
+        Payload::VerifierSet(_) => COMMAND_TYPE_ROTATE_SIGNERS,
+    };
```

Comment:
```suggestion
    let command_type = payload.variant_to_u8();
```
perhaps? You can define a Sui `CommandType` and implement a `From<Payload> for CommandType` that uses `variant_to_u8` internally.

Reply:
Sure

Reply:
I don't know about using that `variant_to_u8` though. That's like coupling the u8 we use in amlifier contracts with the command type thing we use in the Sui contracts. I can make it an enum and do the From trait but I wouldn't wanna use that `variant_to_u8`.

Reply:
Sure, could still wrap inside a From<Payload> for CommandType implementation